### PR TITLE
move methods from NWBBaseType to Container

### LINKED ISSUE #60: Move NWBBaseType functionality to Container
Some of the functionality in NWB core.py is general-purpose and should eventually be extracted into HDMF, including all of NWBBaseType. 

Aside: tracking modified data, if we want to go down that route, should happen in _setter.

### PR DESCRIPTION
**Pull request by bendichter**
---
name: Pull request
about: Create a pull request to help us improve
title: ''
labels: ''
assignees: ''

---

## Motivation

fix #60 

## How to test the behavior?
```
Show here how to reproduce the new behavior (can be a bug fix or a new feature)
```

## Checklist

- [ ] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?
- [ ] Have you ensured the PR description clearly describes problem and the solution?
- [ ] Is your contribution compliant with our coding style ? This can be checked running `flake8` from the source directory.
- [ ] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf/pulls) for the same change?
- [ ] Have you included the relevant issue number using `#XXX` notation where `XXX` is the issue number ? By including "Fix #XXX" you allow GitHub to close the corresponding issue.

### CODE REVIEW

File: src/hdmf/container.py [hunk 1]
```diff
@@ -108,10 +110,175 @@ def parent(self, parent_container):
...
+            v = self.fields[k]
+            template += "  {}: {}\n".format(k, Container.__smart_str(v, 1))
+        return template
+
 
 class Data(Container):
 
-    @abc.abstractproperty
+    @property
+    @abc.abstractmethod
```
Comment (rly) [inline src/hdmf/container.py]: What does this change do?

File: src/hdmf/container.py [hunk 2]
```diff
@@ -124,16 +291,49 @@ def __nonzero__(self):
...
+            key = key.strip()
+            val = val.strip()
+            if key != 'name':
+                ret = list()
+                for item in self.values():
+                    if getattr(item, key, None) == val:
+                        ret.append(item)
+                return ret if len(ret) else None
+            key = val
+        return super(LabelledDict, self).__getitem__(key)
```
Comment (rly) [inline src/hdmf/container.py]: I would move LabelledDict to utils.py

File: src/hdmf/container.py [hunk 3]
```diff
@@ -108,10 +110,175 @@ def parent(self, parent_container):
         else:
             self.__parent = parent_container
 
+    @docval(
+        {'name': 'hdmf_data_type', 'type': str,
+         'doc': 'the hdmf_data_type to search for', 'default': None})
+    def get_ancestor(self, **kwargs):
+        """
+        Traverse parent hierarchy and return first instance of the specified neurodata_type
```
Comment (rly) [inline src/hdmf/container.py]: remove reference to neurodata_type

File: src/hdmf/container.py [hunk 4]
```diff
@@ -108,10 +110,175 @@ def parent(self, parent_container):
...
+        """
+        data_type = getargs('hdmf_data_type', kwargs)
+        if data_type is None:
+            return self.parent
+        p = self.parent
+        while p is not None:
+            if p.hdmf_data_type == data_type:
```
Comment (rly) [inline src/hdmf/container.py]: I don't think `p.hdmf_data_type` resolves...

**Comment by rly** [comment]
`get_ancestor` is used in only one place: In misc.py in the Units class:
```python
nwbfile = self.get_ancestor(neurodata_type='NWBFile')
```

I suggest changing `get_ancestor` to accept a class, e.g. `get_ancestor(type=NWBFile)`, and then the search would be `if isinstance(p, type)`

**Comment by rly** [comment]
@bendichter please also add the changes from https://github.com/NeurodataWithoutBorders/pynwb/pull/998

**Comment by bendichter** [comment]
@rly good to go?

**Comment by bendichter** [comment]
The diff coverage looks bad but the it's all printing that is untested and those tests are in pynwb

**Comment by rly** [comment]
I was waiting to merge until I had written some tests for all these methods. I think some tests should exist independent of PyNWB.

**Comment by rly** [comment]
Superseded by #134

**Pull request closed without merging**